alf/utils/reservoir_sampler.py

```diff
@@ -0,0 +1,153 @@
+    every item with the same probability of K/T. Here we make some changes so that
+    each item t can be kept with a probability roughly in proportional to
+    t^s, where s is an integer speed defined by the user.
+
+    You can use this sampler as a replay buffer. One big difference between this
+    sampler and a uniform replay buffer is that samples overflown by a replay
+    buffer will never be sampled again (probability hard cutting-off).
+
+    Probability computation:
+    - The probability of the t-th sample being selected in the set is (s+1)*K/t
+    - The probability of the t-th sample not being replaced by the sample at time
+      t+m is (1 - (s+1)*K/(t+m) * 1/K) = (t+m-s-1) / (t+m)
+    - The over probability of the t-th sample still being kept in the set at time
+      T is
+
+      (s+1)*K/t * (t+1-s-1)/(t+1) * (t+2-s-1)/(t+2) * ... * (T-s-1)/T
+
+      which is proportional to (t-1)(t-2)..(t-s) / ((T-1)(T-2)...(T-s)) and is
+      roughly proportional to t^s, when t,T >> (s+1)*K
+    """
+
+    def __init__(self,
+                 data_spec: tf.TensorSpec,
+                 K,
+                 speed=0,
+                 name="ReservoirSampler"):
+        """
+        Create a reservoir sampler.
+
+        Args:
+            data_spec (nested TensorSpec): spec for the data item
+            K (int): the size of the reservoir set
+            speed (int): a bigger speed results in a faster decay of old samples.
+                When `speed` is 0, it results in the original reservoir sampling.
+            name (str): name of the sampler
+        """
+        super(ReservoirSampler, self).__init__(
+            data_spec=data_spec, capacity=K, name=name)
+        assert isinstance(K, int) and K > 0
+        assert isinstance(speed, int) and speed >= 0
+        self._s = speed
+        self._K = K
+        self._t = tf.Variable(
+            tf.ones((), tf.float32), trainable=False, name="t")
+
+    def _add_batch(self, batch):
+        """
+        Add a batch of samples to the buffer. When the buffer is not full, add
+        whatever many samples to the buffer until full. Once it's full, the samples
+        in the buffer will be randomly replaced.
+
+        Args:
+            batch (nested Tensor): shape should be [batch_size] + data_spec.shape
+        """
+        batch_size = get_nest_batch_size(batch, tf.int32)
+        buffer_space_left = self._capacity - self._current_size
+        batch_replacing_size = batch_size - buffer_space_left
+
+        # Split the batch into two: the first one (if exists) will be directly
+        # added into the buffer; the second one (if exists) will randomly replace
+        # samples in the buffer
+        def _fill_buffer():
+            data_filling_buffer = tf.nest.map_structure(
+                lambda bat: bat[:buffer_space_left], batch)
+            super(ReservoirSampler, self).add_batch(data_filling_buffer)
```

Comment:
super().add_batch

Reply:
That will report errors about incorrect arguments for super().